### ISSUE TITLE
CopyBinariesToOutDir in PerfView.SupportFiles does I/O at inappropriate times

### ISSUE THREAD

**Issue by rainersigwald** [Microsoft/perfview#198]
The NuGet package `PerfView.SupportFiles` defines a target that copies some package-delivered binaries to the build's output folder:

```csproj
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Target Name="CopyBinariesToOutDir" BeforeTargets="ResolveAssemblyReferences" Condition="('$(DesignTimeBuild)' != 'true')">
	<Copy SourceFiles="$(MSBuildThisFileDirectory)..\lib\native\x86\sd.exe" DestinationFolder="$(OutDir)x86"/>
	<Copy SourceFiles="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.dll" DestinationFolder="$(OutDir)x86"/>
	<Copy SourceFiles="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.Interop.dll" DestinationFolder="$(OutDir)x86"/>
	<Copy SourceFiles="$(MSBuildThisFileDirectory)..\lib\net40\Microsoft.Diagnostics.Tracing.EventSource.dll" DestinationFolder="$(OutDir)"/>
	<Copy SourceFiles="$(MSBuildThisFileDirectory)..\lib\net40\Microsoft.DiagnosticsHub.Packaging.InteropEx.dll" DestinationFolder="$(OutDir)"/>
    </Target>
</Project>
```

This goes against convention by doing disk operations before ResolveAssemblyReference. The `DesignTimeBuild` condition should keep it from firing during most VS operations (I actually came here to suggest adding just that condition), but it evidently doesn't keep it from happening in _enough_ situations.

This is the cause of the VS UI hang reported by @sharwell at https://github.com/Microsoft/msbuild/issues/2100.

**Comment by rainersigwald** [comment]
I tried changing the condition to `'$(BuildingProject)' == 'true'`, which seems to work better. But while I'm here I'll propose a bit of an overhaul in the target.

* Do you need to explicitly copy the `lib/` assemblies? I would expect them to be copied from the package by virtue of being `<Reference>`d in the project.
* The copy operations could be done in bulk to reduce a tiny bit of overhead.
* Unless I'm misunderstanding, this target can run much later in the build, around the same time other copy-to-output-directory targets run.

I think something like this would do the trick:

```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\sd.exe" />
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.dll" />
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.Interop.dll" />
  </ItemGroup>

  <Target Name="CopyBinariesToOutDir" BeforeTargets="CopyFilesToOutputDirectory">
    <Copy SourceFiles="@(PerfViewSupportFilesX86)" DestinationFolder="$(OutDir)x86" />
  </Target>
</Project>
```

**Comment by rainersigwald** [comment]
Shame on me: I wrote a target with poor incremental-build behavior. It should actually be more like:

```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\sd.exe" />
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.dll" />
    <PerfViewSupportFilesX86 Include="$(MSBuildThisFileDirectory)..\lib\native\x86\DiagnosticsHub.Packaging.Interop.dll" />
  </ItemGroup>

  <Target Name="CopyBinariesToOutDir" BeforeTargets="CopyFilesToOutputDirectory">
    <Copy SourceFiles="@(PerfViewSupportFilesX86)"
          DestinationFolder="$(OutDir)x86"
          SkipUnchangedFiles="true" />
  </Target>
</Project>
```

With that, the Copy task will notice when everything is up to date and elide the copy, meaning the cost on incremental builds is just a couple of `stat()`s per file.

**Comment by sharwell** [comment]
Thanks for the help with this @rainersigwald, it is much appreciated. Sorry to steal part of your weekend from you (I didn't mean to but I also didn't want to wait until today and maybe forget some details).

**Comment by vancem** [comment]
Unfortunately I just tried @rainersigwald s solution and it did not work.
The basic issue is that perfView is doing something that is relatively unusual, it is embedding various binaries it needs into the EXE as resources.    This allows PerfView to be deployed as a single EXE (the EXE unpacks itself as part of first startup).   

Because of this it needs these binaries to be present BEFORE the compilation of the PerfView.exe executable.   From the behavior I see, CopyFilesToOutputDirectory is too late.

I have added the "'Condition="('$(DesignTimeBuild)' != 'true')"  clause which will probably avoid the problem.   
  
I am open to better solutions.     Fundamentally all that is needed is to be able to reference the DLLs in question in a  <EmbeddedResource> element in the .csproj file when  building PerfView.exe.   However these files live live in a particular package whose version may vary over time.    The solution that is currently used is to have the PACKAGE copy the files to the output directory, so that PerfView's .csproj can find them there.      It is a bit of a hack.  Those that know more about MSbuild and how nuget packages work with it, can probably do better, and I am happy to take suggestions.

**Comment by sharwell** [comment]
@vancem You can include a .props file in the NuGet package which injects a property into the project containing the location of the NuGet package contents, regardless of where they are.

**Comment by rainersigwald** [comment]
Ah, I see. You're embedding these files from `$(OutDir)` in [PerfView.csproj#L512-L531](https://github.com/Microsoft/perfview/blob/21161014a0536044facae1f1bae691df6598e7e4/src/PerfView/PerfView.csproj#L512-L531).

Could you instead just define the EmbeddedResource items in the package? As @sharwell says this could be in the form of a .props file something like

```xml
<Project>
  <ItemGroup>
    <EmbeddedResource Include="$(MSBuildThisFileDirectory)..\lib\native\x86\sd.exe">
      <Type>Non-Resx</Type>
      <WithCulture>false</WithCulture>
      <LogicalName>.\x86\sd.exe</LogicalName>
    </EmbeddedResource>
... and so on
  </ItemGroup>
</Project>
```

Then the files will be embedded directly from the package download location.

**Comment by sharwell** [comment]
> Could you instead just define the EmbeddedResource items in the package?

Oh, nice call. I used that strategy before. I would link to the file but it's not as good an example because I used `$(SolutionDir)\packages\{package}` instead of `$(MSBuildThisFileDirectory)`.

**Comment by vancem** [comment]
I don't really like the idea that the support files have the <EmeddedResource> element in them.    The job of the support files package is to simply supply binaries.   It should not really care what is done with them (PerfView MIGHT embed them in the EXE but it might not).  

I think @sharwell 's idea of a props file sounds fine.    It would eliminate the need for the targets file completely.  

The nuget package for these support files can be generated from the PerfView GitHub repo.  There are scripts in the src\NugetSupportFiles directory and there are instructions here 

https://github.com/Microsoft/perfview/blob/master/documentation/MakingSupportFilesNugetPackages.md

I think I have the general idea of the solution here, (copying what roslyn did in Roslyn.Common.props) to provide a global ID that points at the base of the support files.   I will play with it tonight.

**Comment by sharwell** [comment]
@vancem Cool. There are two pieces of NuGet magic you'll want to know:

1. The only difference between .props and .targets for the sake of NuGet is where they get included in the project. .props files included at top, .targets at bottom. Since you're talking about defining a property that will be referenced in the project file itself, you'll definitely need a .props file.
2. The name is magic. If (and only if) it has the form [PackageName].props, it will be detected by NuGet when the package is installed and included in the project file. For your package, the path and name of this file within the NuGet package *must* be exactly **build\\PerfView.SupportFiles.props**.

**Comment by vancem** [comment]
Got it.  I have a prototype working.

**Comment by vancem** [comment]
Fixed but #210
